### ISSUE TITLE
Important style in the HighlightCode component 

### ISSUE THREAD

**Issue by radoslawroszkowiak** [swagger-api/swagger-ui#4029]
Hi All,
I'm trying to write an external plugin to highlight the JSON / XML syntax.

In general, that should not be a bid deal, but there's one issue.
In https://github.com/swagger-api/swagger-ui/blob/7f19d791f88cec9ceadb9d17c04ae626c36afd83/src/style/_layout.scss#L609 `!important` color style is used, which sabotages the highlighting attempts.

At the same time, the `highlight` function used in the current `HighlightCode` component sets the inline colors https://github.com/swagger-api/swagger-ui/blob/7f19d791f88cec9ceadb9d17c04ae626c36afd83/src/core/utils.js#L247. These inline colors are ignored because of the `!important` from `_layout.scss`.

What's the rationale? Why do we "highlight" code with this utility function only to override it with `!important` CSS color?

If highlighting is not to be implemented in the main SwaggerUI bundle, is there a point in setting these inline styles in `highlight` function (or invoking it at all)?

This [PR](https://github.com/swagger-api/swagger-ui/pull/3396) (trying to implement code highlighting) removes this "important" `_layout.scss` style, but it has been declined. The comments mention it's a perfect candidate for a plugin.

My idea to write an external plugin was simply to replace the `HighlightCode` component with the one that utilizes `highlight.js` internally. Can you provide me with any guidance on how to override this style from `_layout.scss` in a plugin? Is this feasable at all in a plugin?

| Q                               | A
| ------------------------------- | -------
| Bug or feature request?         |  Bug (kind of, I would say)
| Which Swagger/OpenAPI version?  | not relevant
| Which Swagger-UI version?       | 3.5.0
| How did you install Swagger-UI? | Just copied the needed files from the `dist` folder
| Which browser & version?        | chrome, but no reason to be different on any other
| Which operating system?         | any

### Expected Behavior
No important style affects `HighlightCode` component.
`highlight` utility function does not set inline colors if there are not used.

### Current Behavior
Important color style is set in https://github.com/swagger-api/swagger-ui/blob/7f19d791f88cec9ceadb9d17c04ae626c36afd83/src/style/_layout.scss#L609
Inline (not used) colors are set in https://github.com/swagger-api/swagger-ui/blob/7f19d791f88cec9ceadb9d17c04ae626c36afd83/src/core/utils.js#L162

Important style cannot be overridden. It makes potential extensions creation pretty tricky.

If I'm missing something important here, please let me know and sorry to bother you!

**Comment by shockey** [comment]
Hmm, looks like that !important has been there since 3.0.0, and was modified once in #3769:

https://github.com/swagger-api/swagger-ui/blob/9eae562767af6734479be17237c27b3f06e1fc5f/src/style/_layout.scss#L602

If it's as simple as moving that #fff to the inline style in the highlighter, I don't see why we wouldn't do that.

> Can you provide me with any guidance on how to override this style from _layout.scss in a plugin? Is this feasable at all in a plugin?

The plugin API itself doesn't concern itself with CSS - but you would be able to override the built in CSS, if it weren't for that `!important` statement.

In summary: I'd be happy to see a PR that eliminates the `!important` statement!

**Comment by radoslawroszkowiak** [comment]
Thanks for the quick response. I wonder: is there any point in keeping this highlight function at all?

There are some mentions of microlight (this highlight function is derived from microlight, there's also `microlight` class on `pre` element in `HighlightCode` component).

Should microlight highlighting work or it's just some remains of an idea that was never implemented?
It does not work in my setup and it does not work in the http://petstore.swagger.io/ demo. Should this not working code be kept?

If you plan to use plugins for syntax highlighting (as suggested [here](https://github.com/swagger-api/swagger-ui/pull/3396)), wouldn't it be better to:

1. Rename `HighlightCode` into `Code` component that just display `<pre>` with some simple styles (no `!important` of course) and does not atttempt to highlight anything.
2. Remove the `highlight` function from the utils.
3. State in the docs that it's plugins that should be used for code highlighting.
4. If any of the standalone npm plugins turns out to be good enough, it could be included in the main SwaggerUI bundle plugins.

Thoughts? Do I miss something? In case you believe the above approach would be okay, I think I could prepare some PR (although my knowledge of this project is very limited now).

Thanks!

**Comment by radoslawroszkowiak** [comment]
Hey @shockey, could you please let me know what you think of the above? I think I could prepare either the simple PR removing important color and changing inline color in highlight function or the more complex one as mentioned in points 1 and 2 comment above (removing highlight function completely).

**Comment by shockey** [comment]
@radoslawroszkowiak, the approach you outlined sounds good!

As for the E2E tests - yeah, targeting based on `pre.microlight` is not great. Just removing that class should fix the tests, I'd think, which is fine.

As for putting this in a plugin, I'll quote myself from #3396:

> If you'd like to take another swing at this, I'd be happy to consider a PR that replaces the current code highlighter with a superior one - the deal with this one is that it lives in a weird space between plugins and core. But that's not your fault, we need to push out more guidance on how to build plugins!

I'd be happy to see highlighting in the core product - it's just that the PR we passed on was structured as a plugin within the core repository, which wasn't optimal. So feel free to open a PR that adds it to the core itself, if it's performant and pretty 😄

**Comment by radoslawroszkowiak** [comment]
As for the E2E tests - `microlight` class is not an issue. I think targeting on all the `span` markup created by the `highlight` function is not great (as there's no highlighting in fact and the function is currently useless). I think the `HighlightCode` component (in fact `Code` and not `HighlightCode` in all the honesty) could just render simple `pre` without any inner markup. Anyway, such a change would trigger serious refactoring of E2E tests.

Having this in mind, I think I could try to prepare a PR which will:
1. Remove important style, so that plugins for highlighting are more convenient.
2. Since it's already been approached - use microlight for styling (copy the relevant functionality from microlight.js https://github.com/asvd/microlight/blob/b210d51ae78c396edefbace2714e7da62f965089/microlight.js#L123-L141. Currently it's been altered to set inline white color (which does nothing as nothing is highlighted as a result). Basically it's just a change of a couple of lines in the `highlight` utility function to match the original one.
3. Make highlighting via microlight enabled as as a configurable parameter (with a default as false?).
4. Make response length highlight threshold configurable (with the current value as the default).

I'm personally not a fan of microlight, but I guess it might be better than nothing, especially if not compulsory. It will not increase bundle size. I think in my current projects we would use even microlight highlighting in swagger if it was available. Anyway, point 1 when fixed would open a possibility for plugins using i.e. highlight.js.

@shockey, would such a PR have a chance to be merged?

**Comment by shockey** [comment]
@radoslawroszkowiak, that works for me. I do know that code highlighting can be CPU-intensive, so whether you decide to include microlight.js or something else, performance is important.

Especially considering the performance concerns, a new display config option that controls highlighting sounds good to me as well.

Feel free to ping me here or over email (email in profile) if you decide to pick this up and have questions!

**Comment by webron** [comment]
Closing this as an old ticket while doing some housekeeping. We can reopen it if there's interest though, not saying it's not relevant.
